vllm/v1/engine/async_llm.py

```diff
@@ -144,25 +144,26 @@ async def add_request(
         """Add new request to the AsyncLLM."""
 
         # 1) Create a new output queue for the request.
-        if self.output_processor.is_request_active(request_id):
-            raise ValueError(f"Request id {request_id} already running.")
         queue: asyncio.Queue[RequestOutput] = asyncio.Queue()
 
-        # 2) Convert Input --> Request.
-        request = self.processor.process_inputs(request_id, prompt, params,
-                                                arrival_time, lora_request,
-                                                trace_headers,
-                                                prompt_adapter_request,
-                                                priority)
+        # 2) Fan out child requests (for n>1)
+        parent_req = ParentRequest.from_params(request_id, params)
+        for idx in range(parent_req.n):
+            child_req_id, child_params = parent_req.get_child_info(idx)
```

Comment:
WDTY about adjusting this and making `RequestState.parent_req` `Optional`, so that we won't need `SingularSamplingRequest`? Mainly to reduce the abstractions and make the logic here a bit more transparent.
```suggestion
        parent_req = ParentRequest.from_params(request_id, params) if params.n > 1 else None
        for idx in range(params.n):
            if parent_req:
                request_id, params = parent_req.get_child_info(idx)
```

Reply:
All great points Nick, thanks for pushing on this - I think the end result is better for it :+1: